## NA imputation 

In [1]:
import jusipy
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
import missingno as msno
import copy
from fast_imputation import fast_imputation

#### Loading...

In [2]:
all_points = pd.read_pickle('data/all_points.pkl')

In [3]:
all_points.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22107 entries, 0 to 19999
Data columns (total 6 columns):
lat         22107 non-null object
long        22107 non-null float64
country     22107 non-null object
year        22107 non-null object
positive    22107 non-null int64
random      22107 non-null int64
dtypes: float64(1), int64(2), object(3)
memory usage: 1.2+ MB


In [4]:
A_country = jusipy.country_features.All()

Loading WorldBank(/home/marios/Desktop/Personal/repos/JuSiPy_Asser/jusipy/country_features/data/SD_API_15_DS2_en_csv_v2_10186395.csv)                                                                                                    

In [5]:
country_features = jusipy.country_features.get(all_points, A_country, fuzzy=50)

In [6]:
all_points2 = all_points.drop_duplicates(subset=['country','year'])

In [7]:
all_points2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 776 entries, 0 to 19844
Data columns (total 6 columns):
lat         776 non-null object
long        776 non-null float64
country     776 non-null object
year        776 non-null object
positive    776 non-null int64
random      776 non-null int64
dtypes: float64(1), int64(2), object(3)
memory usage: 42.4+ KB


In [8]:
country_features = jusipy.country_features.get(all_points2, A_country, fuzzy=50)

In [9]:
country_features.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 776 entries, 0 to 19844
Columns: 396 entries, WB_LGAF__WB_LGAF2016_1_1_1 to WB_SD__SP_DYN_LE00_MA_IN
dtypes: float64(280), object(116)
memory usage: 2.4+ MB


-------------------------------------------------------------------------------

### NA percentage (in red)

In [10]:
# somehow unconsciouly it came out like the French flag..

In [11]:
fields = ['WB_ARD__AG_SRF_TOTL_K2', 'WB_SE__WB_NY_GDP_MKTP_KD_ZG', 'WB_ARD__SP_RUR_TOTL', 'WB_ARD__EG_ELC_ACCS_RU_ZS', 'WB_ARD__AG_LND_TOTL_RU_K2',
          'WB_ARD__AG_LND_FRST_K2', 'WB_ARD__AG_LND_ARBL_ZS', 'WB_ARD__AG_LND_AGRI_K2', 'WB_ARD__AG_CON_FERT_ZS','WB_ARD__AG_LND_PRCP_MM','WB_ARD__EN_ATM_METH_AG_KT_CE',
         'WB_ARD__EN_ATM_NOXE_AG_KT_CE','WB_ARD__NV_AGR_TOTL_ZS','WB_ARD__TM_VAL_AGRI_ZS_UN','WB_ARD__TX_VAL_AGRI_ZS_UN','WB_ARD__AG_LND_TOTL_K2',
         'WB_ARD__SL_AGR_EMPL_ZS']
fields_3 = list(country_features.keys())
for x in fields_3:
    df = country_features[x]
    if x in fields:
        print('\033[94m'+x+': '+'\033[92m'+A_country.description[x]+' has '+str(sum(df.isna()))+' NA '+'\033[91m'+str(sum(df.isna())/len(df)))
    else:
        print('\033[94m'+x+': '+'\033[0m'+A_country.description[x]+' has '+str(sum(df.isna()))+' NA '+'\033[91m'+str(sum(df.isna())/len(df)))

WB_LGAF__WB_LGAF2016_1_1_1: Rural land tenure rights (i) recognized and (ii) protected in practice has 751 NA 0.9677835051546392
WB_LGAF__WB_LGAF2016_1_1_2: Customary tenure rights are (i) recognized and (ii) protected in practice has 759 NA 0.9780927835051546
WB_LGAF__WB_LGAF2016_1_1_3: Indigenous rights to land & forest are (i) recognized and (ii) protected in practice has 759 NA 0.9780927835051546
WB_LGAF__WB_LGAF2016_1_1_4: Urban land tenure rights are (i) recognized and (ii) protected in practice has 751 NA 0.9677835051546392
WB_LGAF__WB_LGAF2016_1_2_1: Accessible opportunities for tenure individualization exist has 752 NA 0.9690721649484536
WB_LGAF__WB_LGAF2016_1_2_2: Individual land in rural areas is (i) recorded and (ii) mapped has 751 NA 0.9677835051546392
WB_LGAF__WB_LGAF2016_1_2_3: Individual land in urban areas is (i) recorded and (ii) mapped has 751 NA 0.9677835051546392
WB_LGAF__WB_LGAF2016_1_2_4: The number of illegal land sales is low has 757 NA 0.9755154639175257
WB_LG

In [12]:
fields = ['WB_ARD__AG_SRF_TOTL_K2', 'WB_SE__WB_NY_GDP_MKTP_KD_ZG', 'WB_ARD__SP_RUR_TOTL', 'WB_ARD__EG_ELC_ACCS_RU_ZS', 'WB_ARD__AG_LND_TOTL_RU_K2',
          'WB_ARD__AG_LND_FRST_K2', 'WB_ARD__AG_LND_ARBL_ZS', 'WB_ARD__AG_LND_AGRI_K2', 'WB_ARD__AG_CON_FERT_ZS','WB_ARD__AG_LND_PRCP_MM','WB_ARD__EN_ATM_METH_AG_KT_CE',
         'WB_ARD__EN_ATM_NOXE_AG_KT_CE','WB_ARD__NV_AGR_TOTL_ZS','WB_ARD__TM_VAL_AGRI_ZS_UN','WB_ARD__TX_VAL_AGRI_ZS_UN','WB_ARD__AG_LND_TOTL_K2',
         'WB_ARD__SL_AGR_EMPL_ZS']
imputed_ = []
for x in fields:
    df_temp = country_features[x]
    imputed_.append(fast_imputation(df_temp))

In [13]:
all_points_new = copy.deepcopy(all_points)

#### New dataset

In [14]:
for i in range(len(fields)):
    x = fields[i]
    all_points_new[x] = imputed_[i]

all_points_new.to_pickle('./data/all_points_new.pkl')

In [15]:
all_points_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22107 entries, 0 to 19999
Data columns (total 23 columns):
lat                             22107 non-null object
long                            22107 non-null float64
country                         22107 non-null object
year                            22107 non-null object
positive                        22107 non-null int64
random                          22107 non-null int64
WB_ARD__AG_SRF_TOTL_K2          1525 non-null float64
WB_SE__WB_NY_GDP_MKTP_KD_ZG     1525 non-null float64
WB_ARD__SP_RUR_TOTL             1525 non-null float64
WB_ARD__EG_ELC_ACCS_RU_ZS       1525 non-null float64
WB_ARD__AG_LND_TOTL_RU_K2       1525 non-null float64
WB_ARD__AG_LND_FRST_K2          1525 non-null float64
WB_ARD__AG_LND_ARBL_ZS          1525 non-null float64
WB_ARD__AG_LND_AGRI_K2          1525 non-null float64
WB_ARD__AG_CON_FERT_ZS          1525 non-null float64
WB_ARD__AG_LND_PRCP_MM          1525 non-null float64
WB_ARD__EN_ATM_METH_AG_KT_CE

#### The columns added are (in order):

In [16]:
for x in fields:
    print(A_country.description[x])

Surface area (sq. km)
GDP growth (annual %)
Rural population
Access to electricity, rural (% of rural population)
Rural land area (sq. km)
Forest area (sq. km)
Arable land (% of land area)
Agricultural land (sq. km)
Fertilizer consumption (kilograms per hectare of arable land)
Average precipitation in depth (mm per year)
Agricultural methane emissions (thousand metric tons of CO2 equivalent)
Agricultural nitrous oxide emissions (thousand metric tons of CO2 equivalent)
Agriculture, forestry, and fishing, value added (% of GDP)
Agricultural raw materials imports (% of merchandise imports)
Agricultural raw materials exports (% of merchandise exports)
Land area (sq. km)
Employment in agriculture (% of total employment) (modeled ILO estimate)


#### Also this paper https://academic.oup.com/bioinformatics/article/21/8/1509/249540, suggests that for highly correlated features the optimal feature size is
#### sqrt(N), where N is the sample size. For us this would be roughly 116 features.. 

In [18]:
gdp = country_features['WB_SE__WB_NY_GDP_MKTP_KD_ZG']

4.834865000000001